Breakdown metrics

gregkalapos opened the pull request:
Solves https://github.com/elastic/apm-agent-dotnet/issues/227

![image](https://user-images.githubusercontent.com/1091853/115612915-139d7580-a2ec-11eb-9226-9320b061e9bb.png)


TODOs:
- [x] Add docs
- [x] ~Test with non-sampled transactions~ update: not needed - non-sampled transactions and spans also generate these metrics - according to the spec that's preferred.
- [x] Disable breakdown metric collection

apmmachine commented:
## :grey_exclamation: Build Aborted
  
  > There is a new build on-going so the previous on-going builds have been aborted.
  


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-dotnet%2Fapm-agent-dotnet-mbp%2FPR-1271/detail/PR-1271/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-dotnet%2Fapm-agent-dotnet-mbp%2FPR-1271/detail/PR-1271/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-dotnet%2Fapm-agent-dotnet-mbp%2FPR-1271/detail/PR-1271/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-dotnet%2Fapm-agent-dotnet-mbp%2FPR-1271/detail/PR-1271/1//artifacts) 

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats


* Build Cause: `Pull request #1271 opened`


* Reason: `Aborted from #2`


* Start Time: 2021-04-21T19:46:25.172+0000


* Duration: 13 min 12 sec


* Commit: 7aa09708387735129419b99177f00cbc36af4509




#### Trends :test_tube:


![Image of Build Times](https://apm-ci.elastic.co/job/apm-agent-dotnet/job/apm-agent-dotnet-mbp/job/PR-1271/buildTimeGraph/png)



</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->









### Steps errors [![3](https://img.shields.io/badge/3%20-red)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-dotnet%2Fapm-agent-dotnet-mbp%2FPR-1271/detail/PR-1271/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>


  
  
  
  
##### `Archive the artifacts`
<ul>
<li>Took 0 min 0 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-dotnet/pipelines/apm-agent-dotnet-mbp/pipelines/PR-1271/runs/1/steps/424/log/?start=0">here</a></li>
<li>Description: <code>[2021-04-21T19:59:34.063Z] Archiving artifacts
withAzureCredentials: error org.jenkinsci.plugins.wo</code></l1>
</ul>
  
  
  
  
##### `Build`
<ul>
<li>Took 1 min 36 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-dotnet/pipelines/apm-agent-dotnet-mbp/pipelines/PR-1271/runs/1/steps/289/log/?start=0">here</a></li>
<li>Description: <code>.ci/windows/dotnet.bat</code></l1>
</ul>
  
  
  
  
##### `Error signal`
<ul>
<li>Took 0 min 0 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-dotnet/pipelines/apm-agent-dotnet-mbp/pipelines/PR-1271/runs/1/steps/324/log/?start=0">here</a></li>
<li>Description: <code>withAzureCredentials: error org.jenkinsci.plugins.workflow.steps.FlowInterruptedException</code></l1>
</ul>
  
  </p>
  </details>



<!-- LOG OUTPUT-->
### Log output

<details><summary>Expand to view the last 100 lines of log output</summary>
<p>

```
[2021-04-21T19:59:22.489Z]              This reference is not "CopyLocal" because at least one source item had "Private" set to "false" and no source items had "Private" set to "true".
[2021-04-21T19:59:22.489Z]   14:8>ResolveAssemblyReferences:
[2021-04-21T19:59:22.489Z]          Primary reference "Microsoft.AspNetCore.Session, Version=2.1.1.0, Culture=neutral, PublicKeyToken=adb9793829ddae60".
[2021-04-21T19:59:22.489Z]   41:6>ResolveAssemblyReferences:
[2021-04-21T19:59:22.489Z]              The ImageRuntimeVersion for this reference is "v4.0.30319".
[2021-04-21T19:59:22.489Z]   14:8>ResolveAssemblyReferences:
[2021-04-21T19:59:22.489Z]              Resolved file path is "C:\Program Files\dotnet\sdk\NuGetFallbackFolder\microsoft.aspnetcore.session\2.1.1\lib\netstandard2.0\Microsoft.AspNetCore.Session.dll".
[2021-04-21T19:59:22.489Z]   41:6>ResolveAssemblyReferences:
[2021-04-21T19:59:22.489Z]          Primary reference "System.Linq.Expressions, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a".
[2021-04-21T19:59:22.489Z]   14:8>ResolveAssemblyReferences:
[2021-04-21T19:59:22.489Z]              Reference found at search path location "{HintPathFromItem}".
[2021-04-21T19:59:22.489Z]   41:6>ResolveAssemblyReferences:
[2021-04-21T19:59:22.489Z]              Resolved file path is "C:\Program Files\dotnet\packs\Microsoft.NETCore.App.Ref\5.0.0\ref\net5.0\System.Linq.Expressions.dll".
[2021-04-21T19:59:22.489Z]   14:8>ResolveAssemblyReferences:
[2021-04-21T19:59:22.489Z]              This reference is not "CopyLocal" because at least one source item had "Private" set to "false" and no source items had "Private" set to "true".
[2021-04-21T19:59:22.489Z]   41:6>ResolveAssemblyReferences:
[2021-04-21T19:59:22.489Z]              Reference found at search path location "{RawFileName}".
[2021-04-21T19:59:22.489Z]              This reference is not "CopyLocal" because at least one source item had "Private" set to "false" and no source items had "Private" set to "true".
[2021-04-21T19:59:22.489Z]   14:8>ResolveAssemblyReferences:
[2021-04-21T19:59:22.489Z]              The ImageRuntimeVersion for this reference is "v4.0.30319".
[2021-04-21T19:59:22.489Z]   41:6>ResolveAssemblyReferences:
[2021-04-21T19:59:22.489Z]              The ImageRuntimeVersion for this reference is "v4.0.30319".
[2021-04-21T19:59:22.489Z]   14:8>ResolveAssemblyReferences:
[2021-04-21T19:59:22.489Z]          Primary reference "Microsoft.AspNetCore.SignalR, Version=1.0.1.0, Culture=neutral, PublicKeyToken=adb9793829ddae60".
[2021-04-21T19:59:22.489Z]   41:6>ResolveAssemblyReferences:
[2021-04-21T19:59:22.489Z]          Primary reference "System.Linq.Parallel, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a".
[2021-04-21T19:59:22.490Z]   14:8>ResolveAssemblyReferences:
[2021-04-21T19:59:22.490Z]              Resolved file path is "C:\Program Files\dotnet\sdk\NuGetFallbackFolder\microsoft.aspnetcore.signalr\1.0.1\lib\netstandard2.0\Microsoft.AspNetCore.SignalR.dll".
[2021-04-21T19:59:22.490Z]   41:6>ResolveAssemblyReferences:
[2021-04-21T19:59:22.490Z]              Resolved file path is "C:\Program Files\dotnet\packs\Microsoft.NETCore.App.Ref\5.0.0\ref\net5.0\System.Linq.Parallel.dll".
[2021-04-21T19:59:22.495Z] script returned exit code -1
[2021-04-21T19:59:22.935Z] Downloading visualstudio2019professional 
[2021-04-21T19:59:22.935Z]   from 'https://download.visualstudio.microsoft.com/download/pr/5c555b0d-fffd-45a2-9929-4a5bb59479a4/c3654155226ad307bc7d7da61e28aff50742ccb033c178debed815d92137fdb9/vs_Professional.exe'
[2021-04-21T19:59:22.935Z] 
[2021-04-21T19:59:22.935Z] Download of vs_Professional.exe (1.39 MB) completed.
[2021-04-21T19:59:23.051Z]        _GetAllRestoreProjectPathItems:
[2021-04-21T19:59:23.052Z]          Determining projects to restore...
[2021-04-21T19:59:23.083Z] Sending interrupt signal to process
[2021-04-21T19:59:25.948Z] Sending interrupt signal to process
[2021-04-21T19:59:26.246Z] script returned exit code -1
[2021-04-21T19:59:27.020Z] Stage "Build - MSBuild" skipped due to earlier failure(s)
[2021-04-21T19:59:27.135Z] Stage "Test" skipped due to earlier failure(s)
[2021-04-21T19:59:27.257Z] Stage "IIS Tests" skipped due to earlier failure(s)
[2021-04-21T19:59:27.262Z] Terminated
[2021-04-21T19:59:27.270Z] script returned exit code 143
[2021-04-21T19:59:27.351Z] Post stage
[2021-04-21T19:59:27.384Z] Running in /var/lib/jenkins/workspace/net_apm-agent-dotnet-mbp_PR-1271/apm-agent-dotnet
[2021-04-21T19:59:27.471Z] [WS-CLEANUP] Deleting project workspace...
[2021-04-21T19:59:27.471Z] [WS-CLEANUP] Deferred wipeout is disabled by the job configuration...
[2021-04-21T19:59:27.894Z] + rm .credentials.json
[2021-04-21T19:59:27.988Z] $ docker stop --time=1 9fd1848c380e8dae8745c0d2f6228211a49603bf81de4eec86af34f9a77aefa0
[2021-04-21T19:59:28.833Z] [WS-CLEANUP] done
[2021-04-21T19:59:29.413Z] $ docker rm -f 9fd1848c380e8dae8745c0d2f6228211a49603bf81de4eec86af34f9a77aefa0
[2021-04-21T19:59:29.650Z] Post stage
[2021-04-21T19:59:29.827Z] Archiving artifacts
[2021-04-21T19:59:29.980Z] Failed in branch Windows .NET Framework
[2021-04-21T19:59:30.146Z] Stage "Test" skipped due to earlier failure(s)
[2021-04-21T19:59:30.273Z] Stage "Startup Hook Tests" skipped due to earlier failure(s)
[2021-04-21T19:59:30.355Z] Post stage
[2021-04-21T19:59:30.383Z] [WS-CLEANUP] Deleting project workspace...
[2021-04-21T19:59:30.383Z] [WS-CLEANUP] Deferred wipeout is disabled by the job configuration...
[2021-04-21T19:59:30.578Z] Post stage
[2021-04-21T19:59:30.606Z] Running in /var/lib/jenkins/workspace/net_apm-agent-dotnet-mbp_PR-1271/apm-agent-dotnet
[2021-04-21T19:59:30.663Z] Archiving artifacts
[2021-04-21T19:59:30.782Z] Recording test results
[2021-04-21T19:59:31.008Z] None of the test reports contained any result
[2021-04-21T19:59:31.009Z] [Checks API] No suitable checks publisher found.
[2021-04-21T19:59:31.100Z] Publishing Coverage report....
[2021-04-21T19:59:31.170Z] No reports were found
[2021-04-21T19:59:31.279Z] [INFO] getVaultSecret: Getting secrets
[2021-04-21T19:59:31.581Z] Masking supported pattern matches of $VAULT_ADDR or $VAULT_ROLE_ID or $VAULT_SECRET_ID
[2021-04-21T19:59:32.212Z] Running in /var/lib/jenkins/workspace/net_apm-agent-dotnet-mbp_PR-1271/apm-agent-dotnet
[2021-04-21T19:59:32.261Z] [INFO] Codecov: Getting branch ref...
[2021-04-21T19:59:32.478Z] Masking supported pattern matches of $GITHUB_TOKEN
[2021-04-21T19:59:32.942Z] [INFO] Codecov: Sending data...
[2021-04-21T19:59:32.943Z] [WS-CLEANUP] done
[2021-04-21T19:59:33.413Z] Failed in branch Windows .NET Core
[2021-04-21T19:59:33.450Z] + echo curl -sSLo codecov.sh https://codecov.io/bash
[2021-04-21T19:59:33.450Z] curl -sSLo codecov.sh https://codecov.io/bash
[2021-04-21T19:59:33.821Z] + echo bash codecov.sh
[2021-04-21T19:59:33.821Z] bash codecov.sh
[2021-04-21T19:59:34.063Z] Archiving artifacts
[2021-04-21T19:59:34.336Z] Failed in branch Linux
[2021-04-21T19:59:34.702Z] Stage "Release to feedz.io" skipped due to earlier failure(s)
[2021-04-21T19:59:34.759Z] Click here to forcibly terminate running steps
[2021-04-21T19:59:34.817Z] Stage "Release" skipped due to earlier failure(s)
[2021-04-21T19:59:34.868Z] Stage "Release" skipped due to earlier failure(s)
[2021-04-21T19:59:34.990Z] Stage "Release" skipped due to earlier failure(s)
[2021-04-21T19:59:35.188Z] Stage "AfterRelease" skipped due to earlier failure(s)
[2021-04-21T19:59:35.240Z] Stage "AfterRelease" skipped due to earlier failure(s)
[2021-04-21T19:59:35.662Z] Running on Jenkins in /var/lib/jenkins/workspace/net_apm-agent-dotnet-mbp_PR-1271
[2021-04-21T19:59:35.774Z] [INFO] getVaultSecret: Getting secrets
[2021-04-21T19:59:36.083Z] Masking supported pattern matches of $VAULT_ADDR or $VAULT_ROLE_ID or $VAULT_SECRET_ID
[2021-04-21T19:59:37.180Z] + chmod 755 generate-build-data.sh
[2021-04-21T19:59:37.180Z] + ./generate-build-data.sh https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-dotnet/apm-agent-dotnet-mbp/PR-1271/ https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-dotnet/apm-agent-dotnet-mbp/PR-1271/runs/1 ABORTED 791735
[2021-04-21T19:59:37.731Z] INFO: curl https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-dotnet/apm-agent-dotnet-mbp/PR-1271/runs/1/steps/?limit=10000 -o steps-info.json
[2021-04-21T19:59:38.281Z] INFO: curl https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-dotnet/apm-agent-dotnet-mbp/PR-1271/runs/1/tests/?status=FAILED -o tests-errors.json
[2021-04-21T19:59:38.282Z] Retry 1/3 exited 22, retrying in 1 seconds...
[2021-04-21T19:59:39.625Z] Retry 2/3 exited 22, retrying in 2 seconds...
[2021-04-21T19:59:41.486Z] Retry 3/3 exited 22, no more retries left.

```

</p>
</details>



<!--COMMENT_GENERATED_WITH_ID_comment.id-->

bluetentacle commented:
What's the status of this PR?

gregkalapos commented on docs/metrics.asciidoc:
```diff
@@ -174,4 +174,70 @@ format: ms
+
+[float]
+[[metrics-application]]
+=== Built-in application metrics
```
100% stolen from the java docs.

gregkalapos commented:
> What's the status of this PR?

Ready for review.

gregkalapos commented:
Pushed a few small changes:
- After asking around I learned that the duration of spans with `type: "app"` needs to be handled as part of the transaction self-time duration, I adapted the code according to this and added tests.
- Rebased on master.

russcam commented on src/Elastic.Apm/Model/Transaction.cs:
```diff
@@ -477,6 +504,13 @@ public void End()
+			else
+				SpanTimings.TryAdd(("app", null), new SpanTimer(SelfDuration));
+
+			_breakdownMetricsProvider?.CaptureTransaction(this);
```
Could breakdown metrics be captured by subscribing to the `Transaction.End` event in `Tracer`? It would avoid the need to pass the breakdown metrics provider to each transaction to call in the `End()` method.

gregkalapos replied:
Yes, we can do that. In addition to that we also don't need to pass `_breakdownMetricsProvider` to spans - so I also removed it from the span `.ctor`.

russcam reviewed (commented):
Left some initial comments to discuss

gregkalapos commented on docs/metrics.asciidoc:
```diff
@@ -174,4 +174,70 @@ format: ms
+
+Fields:
+
+* `sum.us`: The sum of all transaction durations in ms since the last report (the delta)
```
Hahh - I looked into this and indeed there was an issue here. 

>  is the `u` in `.us` intended to be `μ` (mu)?

Yes! 

> but the description says ms, milliseconds. Is that correct?

No, it wasn't. This was a copy from the java agent docs, and in fact we sent the data in milliseconds in the PR originally. I looked into more docs and this clearly should be microseconds (e.g. the Go agent docs has this correctly). I fixed the code, so now we send it in microseconds and I also fixed this part of the doc as well - all microseconds now.

Side-note: since the breakdown metrics are used to calculate the time spent in different span types by percentage, I think the unit will not break the feature as long as all the data is in the same unit (which was the case), so likely that's why it was not an issue before - it'd be nice to see if all agents send it in the right unit :)

gregkalapos commented on docs/metrics.asciidoc:
```diff
@@ -174,4 +174,70 @@ format: ms
+
+Fields:
+
+* `sum.us`: The sum of all span self-times in ms since the last report (the delta)
```
Same as above.

gregkalapos commented:
Rebased, ready to another review round.

russcam commented on src/Elastic.Apm/Metrics/MetricsCollector.cs:
```diff
@@ -64,6 +66,19 @@ public MetricsCollector(IApmLogger logger, IPayloadSender payloadSender, IConfig
+				{
+					if (item != null)
+					{
+						if (!WildcardMatcher.IsAnyMatch(currentConfigSnapshot.DisableMetrics, BreakdownMetricsProvider.SpanSelfTime))
```
This makes the assumption that any added metrics providers should be controlled by the span self-time metric. Whilst this holds true now as breakdown provider is the only one added in the ctor, maybe the string(s) to match should be a property on an `IMetricsProvider` to allow each individual provider determine if it should be added.

gregkalapos replied:
Yeah, I agree that this isn't ok.

> maybe the string(s) to match should be a property on an IMetricsProvider to allow each individual provider determine if it should be added.

Problem is we'd need to address 2 scenarios with this:
- The `BreakdownMetricsProvider` and everything passed to the `ctor` is already created - so yes, in that case we could just query the string(s) from `IMetricsProvider` and it'd indeed help.
- Every other `IMetricsProvider` is not instantiated at this point, so we can't query the string(s) list for those. In that cases what we do is that we check the `DisabledMetrics` list and if one of those is disabled we don't even instantiate those.

What I did is that I moved the check to `AgentCompontents` - that's where we instantiate the `BreakdownMetricsProvider`. 

Long term we could come up with an infrastructure to better handle this - I think it'd be fine to introduce the string list you suggest, but that'll also mean we'd need to instantiate all the `IMetricsProvider` and if they are not enabled based on the string(s) list, we'd just not add them to the list and let them GCed. Nevertheless that'd be a longer discussion, so maybe we could do that in a separate PR.

gregkalapos replied:
Update: I decided to move on and implement it nevertheless. I added `IMetricsProvider.IsEnabled` - so we instantiate each `IMetricsProvider` implementation, but only add them if they are enabled. This changed a few things around metrics collection, but in my opinion this is now much cleaner.

russcam replied:
++ it'd be possible to build on this for the custom metrics scenarios.

russcam reviewed (commented):
I've gone through in some more detail.

The gathering of breakdown metrics looks like it could be performed in `PayloadSenderV2`, when the transaction is queued. `PayloadSenderV2` would take `BreakdownMetricsProvider` in its ctor. I know I previously suggested hooking into the `.Ended` event as has been done here, but in retrospect, performing in `PayloadSenderV2` would actually be cleaner.

An aside- 

`PayloadSenderV2` is a little bit of a misnomer as its performing a few different functions: 

1. holding a queue of items to send to APM server
2. filtering items that should be sent
3. serializing items to be sent
4. actually sending items to APM server

Might be useful to refactor this in the next major version.

gregkalapos commented on src/Elastic.Apm/Model/Transaction.cs:
```diff
@@ -13,25 +15,35 @@
+
+		internal readonly TraceState _traceState;
+
+		internal readonly ConcurrentDictionary<(string, string), SpanTimer> SpanTimings = new();
```
Done

gregkalapos commented:
> I've gone through in some more detail.
> 
> The gathering of breakdown metrics looks like it could be performed in `PayloadSenderV2`, when the transaction is queued. `PayloadSenderV2` would take `BreakdownMetricsProvider` in its ctor. I know I previously suggested hooking into the `.Ended` event as has been done here, but in retrospect, performing in `PayloadSenderV2` would actually be cleaner.
> 

There are multiple places where we can capture the metrics - basically anything works after the transaction ends and before it goes out to APM server. 

In what sense do you think it would be cleaner? 

I see a few issues with that:
- `PayloadSenderV2` is not a component to create metrics, it's to send the payload.
- If someone replaces it (which is super rare, but I saw some people doing it), they won't have breakdown metrics anymore - which to me makes putting it into `PayloadSenderV2` not clean. If I replace the payload sender, I'd not expect that affecting breakdown metric collection. Same applies for us: let's say we implement a new payload sender - why does it need to change breakdown metrics collection as well?
- The breakdown metrics are emitted by transactions - it's defined to be emitted when the transaction finishes, so I think the closer it is to `Transaction.End()`, the better - yes, `Transaction.EndI()` directly calls `PayloadSenderV2`, but that is a different component for different purposes.
- If transactions emit the metrics, testing is also very natural - no additional config is needed. If we move it into `PayloadSenderV2`, we'll need to inject a `BreakdownMetricsProvider` into our test payload sender. This is not the end of the world, but again raises the question: why does the payload sender influence breakdown metrics collection, which is a metric purely based on transactions (and spans)?

I opened https://github.com/elastic/apm-agent-dotnet/pull/1341 which moves `BreakdownMetricsProvider` to `PayloadSenderV2`. Overall I think the PR in its current form is better. If it's ok for you I'd leave it out, if not I'll cherry pick the commit that moved `BreakdownMetricsProvider` to `PaylaadSenderV2`.

Otherwise it's ready for another round.

russcam commented:
> In what sense do you think it would be cleaner?

1. It avoids the creation and invocation of an event delegate for each transaction, to capture breakdown metrics
2. Breakdown metrics _could_ be captured on the dedicated payload processing thread

The `transaction.Ended` event has a valid use case in scenarios where some integration outside of Elastic.Apm needs to be notified when a transaction ends, such as in StackExchange.Redis, but doesn't have to be used for breakdown metrics, which are calculated in Elastic.Apm.

> * `PayloadSenderV2` is not a component to create metrics, it's to send the payload.

It's doing a lot more than sending payloads; it's more of a processing pipeline for payloads, which could also include calculating breakdown metrics for a particular payload, transactions.

> * If someone replaces it (which is super rare, but I saw some people doing it), they won't have breakdown metrics anymore - which to me makes putting it into `PayloadSenderV2` not clean. If I replace the payload sender, I'd not expect that affecting breakdown metric collection. Same applies for us: let's say we implement a new payload sender - why does it need to change breakdown metrics collection as well?

If someone replaces `IPayloadSender`, the same holds true for 

1. transaction, span and error filtering
2. enqueuing payloads and processing on a configurable interval on a dedicated thread
3. sending payloads to APM server

but the public interface doesn't define a contract for this behaviour, only that payloads are queued

```c#
public interface IPayloadSender
{
	void QueueError(IError error);
	void QueueMetrics(IMetricSet metrics);
	void QueueSpan(ISpan span);
	void QueueTransaction(ITransaction transaction);
}
```

I'm trying to think about the architecture of the agent in terms of how it is evolving and where we might want to go with it. Happy to discuss here, though it might be easier to chat over zoom 😃

russcam commented on src/Elastic.Apm/Metrics/MetricsProvider/BreakdownMetricsProvider.cs:
```diff
@@ -0,0 +1,87 @@
+
+		public IEnumerable<MetricSet> GetSamples()
+		{
+			var retVal = new List<MetricSet>();
```
could avoid resizing the internal array by allocating a capacity based on count of `_itemsToSend` (or 1000)

gregkalapos replied:
Yes, makes sense - done.

russcam commented on src/Elastic.Apm/Metrics/MetricsProvider/BreakdownMetricsProvider.cs:
```diff
@@ -0,0 +1,87 @@
+			lock (_lock)
+			{
+				retVal.AddRange(_itemsToSend.Take(1000));
+				_itemsToSend.Clear();
```
should `_itemsToSend` be cleared? What if it still contains items to send after taking 1000?

gregkalapos replied:
> should _itemsToSend be cleared? What if it still contains items to send after taking 1000?

The spec says: "...agents must cap the total number of metricsets. A reasonable default seems to be  [1000]. When the limit is reached, agents should log sla warning...". So my understanding is that everything above 1k should be thrown away, so it's fine to clear those. Also worth noting that in this case we'd try to show time spent in percentage across 1K groups, so at this point I'd say the visualization can be also challenging.

On the other hand by thinking about this: we should not even collect those in `_itemsToSend`, plus the warning was also missing. So I moved the handling of this into `CaptureTransaction(Transaction)`: we only collect up to 1K items and if 1K is reached we log.

I also included a test to cover this.

russcam commented on src/Elastic.Apm/Model/SpanTypeAndSubtype.cs:
```diff
@@ -0,0 +1,19 @@
+	/// <summary>
+	/// Encapsulates type and subtype
+	/// </summary>
+	internal struct SpanTypeAndSubtype
```
minor nitpick: naming

```suggestion
	internal struct SpanTimerKey
```

gregkalapos replied:
Done.

gregkalapos commented:
> > In what sense do you think it would be cleaner?
> 
> 1. It avoids the creation and invocation of an event delegate for each transaction, to capture breakdown metrics
> 2. Breakdown metrics _could_ be captured on the dedicated payload processing thread
> 
> The `transaction.Ended` event has a valid use case in scenarios where some integration outside of Elastic.Apm needs to be notified when a transaction ends, such as in StackExchange.Redis, but doesn't have to be used for breakdown metrics, which are calculated in Elastic.Apm.

1. I agree the delegate is not optimal for that. So far we had 3 options: 1) doing it in transaction 2) doing it in tracer 3) doing it in payload sender. Option 1 (the original code) would not have this issue and I think it's more natural.
2. That's true for creating the `MetricSet`, but not true for all the calculation. Also, this'd only apply to this single metric, but not for the other. If this is a goal (which I think in this PR should not be) then we should make a dedicated thread for all metrics collection.

> 
> > * `PayloadSenderV2` is not a component to create metrics, it's to send the payload.
> 
> It's doing a lot more than sending payloads; it's more of a processing pipeline for payloads, which could also include calculating breakdown metrics for a particular payload, transactions.
> 

That's true and everyone (rightfully) criticises it, yet we end up putting more stuff into it. So this is again a reason why I'd not put it into there - but the main reason I would not put it there is that I think breakdown metrics creation is really not the job for the payload sender.

> > * If someone replaces it (which is super rare, but I saw some people doing it), they won't have breakdown metrics anymore - which to me makes putting it into `PayloadSenderV2` not clean. If I replace the payload sender, I'd not expect that affecting breakdown metric collection. Same applies for us: let's say we implement a new payload sender - why does it need to change breakdown metrics collection as well?
> 
> If someone replaces `IPayloadSender`, the same holds true for
> 
> 1. transaction, span and error filtering
> 2. enqueuing payloads and processing on a configurable interval on a dedicated thread
> 3. sending payloads to APM server
> 
> but the public interface doesn't define a contract for this behaviour, only that payloads are queued
> 
> ```cs
> public interface IPayloadSender
> {
> 	void QueueError(IError error);
> 	void QueueMetrics(IMetricSet metrics);
> 	void QueueSpan(ISpan span);
> 	void QueueTransaction(ITransaction transaction);
> }
> ```
 
That's true and unfortunate, but why do we need to make yet another component part of this list? I still don't see how it's related to payload sender. Breakdown is a metric created by spans and transactions.

> I'm trying to think about the architecture of the agent in terms of how it is evolving and where we might want to go with it. Happy to discuss here, though it might be easier to chat over zoom 😃

Ok, let's zoom - maybe I just don't get the core of your motivation. I think the cleanest and simplest was to just call `BreakdownMetricsProvider?.CaptureTransaction(Transaction)` from `Transaction.End()` right after we create the last breakdown metric for the transaction. And I feel that holds true regardless of how future payloadsenders may look. But let's zoom...

gregkalapos commented:
> > > In what sense do you think it would be cleaner?
> > 
> > 
> > 
> > 1. It avoids the creation and invocation of an event delegate for each transaction, to capture breakdown metrics
> > 2. Breakdown metrics _could_ be captured on the dedicated payload processing thread
> > 
> > The `transaction.Ended` event has a valid use case in scenarios where some integration outside of Elastic.Apm needs to be notified when a transaction ends, such as in StackExchange.Redis, but doesn't have to be used for breakdown metrics, which are calculated in Elastic.Apm.
> 
> 1. I agree the delegate is not optimal for that. So far we had 3 options: 1) doing it in transaction 2) doing it in tracer 3) doing it in payload sender. Option 1 (the original code) would not have this issue and I think it's more natural.
> 2. That's true for creating the `MetricSet`, but not true for all the calculation. Also, this'd only apply to this single metric, but not for the other. If this is a goal (which I think in this PR should not be) then we should make a dedicated thread for all metrics collection.
> 
> > > * `PayloadSenderV2` is not a component to create metrics, it's to send the payload.
> > 
> > 
> > It's doing a lot more than sending payloads; it's more of a processing pipeline for payloads, which could also include calculating breakdown metrics for a particular payload, transactions.
> 
> That's true and everyone (rightfully) criticises it, yet we end up putting more stuff into it. So this is again a reason why I'd not put it into there - but the main reason I would not put it there is that I think breakdown metrics creation is really not the job for the payload sender.
> 
> > > * If someone replaces it (which is super rare, but I saw some people doing it), they won't have breakdown metrics anymore - which to me makes putting it into `PayloadSenderV2` not clean. If I replace the payload sender, I'd not expect that affecting breakdown metric collection. Same applies for us: let's say we implement a new payload sender - why does it need to change breakdown metrics collection as well?
> > 
> > 
> > If someone replaces `IPayloadSender`, the same holds true for
> > 
> > 1. transaction, span and error filtering
> > 2. enqueuing payloads and processing on a configurable interval on a dedicated thread
> > 3. sending payloads to APM server
> > 
> > but the public interface doesn't define a contract for this behaviour, only that payloads are queued
> > ```cs
> > public interface IPayloadSender
> > {
> > 	void QueueError(IError error);
> > 	void QueueMetrics(IMetricSet metrics);
> > 	void QueueSpan(ISpan span);
> > 	void QueueTransaction(ITransaction transaction);
> > }
> > ```
> 
> That's true and unfortunate, but why do we need to make yet another component part of this list? I still don't see how it's related to payload sender. Breakdown is a metric created by spans and transactions.
> 
> > I'm trying to think about the architecture of the agent in terms of how it is evolving and where we might want to go with it. Happy to discuss here, though it might be easier to chat over zoom 😃
> 
> Ok, let's zoom - maybe I just don't get the core of your motivation. I think the cleanest and simplest was to just call `BreakdownMetricsProvider?.CaptureTransaction(Transaction)` from `Transaction.End()` right after we create the last breakdown metric for the transaction. And I feel that holds true regardless of how future payloadsenders may look. But let's zoom...

Zoom summary: we agreed to move `BreakdownMetricsProvider.CaptureTransaction()` to `Transaction.cs` - I implemented that.

gregkalapos commented on src/Elastic.Apm/Api/Tracer.cs:
```diff
@@ -61,16 +65,35 @@ IApmServerInfo apmServerInfo
 			_logger.Debug()?.Log("Starting {TransactionValue}", retVal);
 			return retVal;
+
+			void RetValOnEnded(object sender, EventArgs e)
```
This part could got removed.

bmorelli25 commented on docs/metrics.asciidoc:
```diff
@@ -174,4 +174,70 @@ format: ms
 Platform: all.
 
 The approximate accumulated collection elapsed time in milliseconds.
+
```
This will fix the doc build:

```suggestion
--

```

gregkalapos commented on test/Elastic.Apm.Tests.Utilities/MockPayloadSender.cs:
```diff
@@ -20,7 +20,10 @@ internal class MockPayloadSender : IPayloadSender
 		private readonly List<IError> _errors = new List<IError>();
 		private readonly List<Func<IError, IError>> _errorFilters = new List<Func<IError, IError>>();
-		private readonly object _lock = new object();
+		private readonly object _spanLock = new object();
```
I saw tests failing with such logs - this specific test was a new one which I introduced in the PR. I wasn't able to repro this locally:

```
System.ArgumentException : Destination array was not long enough. Check the destination index, length, and the array's lower bounds.
Parameter name: destinationArray
Stacktrace
System.ArgumentException : Destination array was not long enough. Check the destination index, length, and the array's lower bounds.
Parameter name: destinationArray
Stack Trace:
   at System.Array.Copy(Array sourceArray, Int32 sourceIndex, Array destinationArray, Int32 destinationIndex, Int32 length, Boolean reliable)
   at System.Collections.Generic.List`1.CopyTo(T[] array, Int32 arrayIndex)
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at Elastic.Apm.Tests.BreakdownTests.TransactionWithSpans() in C:\Users\jenkins\workspace\net_apm-agent-dotnet-mbp_PR-1271\apm-agent-dotnet\test\Elastic.Apm.Tests\BreakdownTests.cs:line 62
   ```
   
  This exception isn't new - I already saw it on failing tests. I think the root cause of those failing tests is that the agent still puts an event into the `MockPayloadSender` while it makes the copy of the list. To address this I introduced locks in the `MockPayloadSender`.

russcam commented on test/Elastic.Apm.Tests.Utilities/MockPayloadSender.cs:
```diff
@@ -72,16 +69,22 @@ public bool WaitForTransactions(TimeSpan? timeout = null, int? count = null)
 				{
 					var stopWatch = Stopwatch.StartNew();
-					while (_transactions.Count < count && signalled)
+					lock (_transactionLock)
```
If I understand this correctly, the lock around waiting for `x` count of transactions can't succeed because adding transactions is also within a lock, meaning no transactions can be added whilst waiting for the count to be reached.

I _think_ the locks around `WaitForX()` count can be removed, and that locking is needed only on adding items or copying to an immutable collection.

gregkalapos replied:
Oh, there is an `AutoResetEvent.WaitOne()` within the lock 🤦 . I really wonder why no test failed, everything was green both locally and in CI. Yeah, I also think this should fail if first thread calls `WaitForX()` and then another one tries to add to the queue.

I think locking on the reading of `_transaction.Count` and `spans.Count` is needed - otherwise the number which the thread calling `WaitForX()` sees can be incorrect if another thread adds to the queue, but the lock should be limited to the call to `.Count`, and that finishes immediately.

I updated the PR accordingly.

russcam reviewed (commented):
Left a comment around locking, rest looks OK.

apmmachine commented:
## :broken_heart: Tests Failed


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-dotnet%2Fapm-agent-dotnet-mbp%2FPR-1271/detail/PR-1271/51//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-dotnet%2Fapm-agent-dotnet-mbp%2FPR-1271/detail/PR-1271/51//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-dotnet%2Fapm-agent-dotnet-mbp%2FPR-1271/detail/PR-1271/51//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-dotnet%2Fapm-agent-dotnet-mbp%2FPR-1271/detail/PR-1271/51//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://apm-agent-dotnet_1271.docs-preview.app.elstc.co/diff)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats


* Build Cause: `Pull request #1271 updated`



* Start Time: 2021-06-23T23:32:33.235+0000


* Duration: 50 min 51 sec


* Commit: 520d06f38e81f908316329612d6d08a058c36964



#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 2  |
| Passed       | 19516  |
| Skipped      | 122 |
| Total        | 19640   |


#### Trends :test_tube:


![Image of Build Times](https://apm-ci.elastic.co/job/apm-agent-dotnet/job/apm-agent-dotnet-mbp/job/PR-1271/buildTimeGraph/png)


![Image of Tests](https://apm-ci.elastic.co/job/apm-agent-dotnet/job/apm-agent-dotnet-mbp/job/PR-1271/51/testReport/history/countGraph/png?start=0&amp;end=25)


</p>
</details>

<!-- TEST RESULTS IF ANY-->

  ### Test errors [![2](https://img.shields.io/badge/2%20-red)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-dotnet%2Fapm-agent-dotnet-mbp%2FPR-1271/detail/PR-1271/51//tests)
<details><summary>Expand to view the tests failures</summary><p>
    
    
##### `Initializing / Parallel / Windows .NET Framework / Test / CentralConfigFetcherTests.Elastic.Apm.Tests.BackendCommTests.CentralConfig.CentralConfigFetcherTests.Should_Update_IgnoreMessageQueues_Configuration – Elastic.Apm.Tests.BackendCommTests.CentralConfig.CentralConfigFetcherTests`


<ul>
<details><summary>Expand to view the error details</summary><p>

```
 Expected collection not to be empty. 
 ```
</p></details>
<details><summary>Expand to view the stacktrace</summary><p>

```
 Expected collection not to be empty.
Stack Trace:
   at FluentAssertions.Execution.XUnit2TestFramework.Throw(String message)
   at FluentAssertions.Execution.TestFrameworkProvider.Throw(String message)
   at FluentAssertions.Execution.DefaultAssertionStrategy.HandleFailure(String message)
   at FluentAssertions.Execution.AssertionScope.FailWith(Func`1 failReasonFunc)
   at FluentAssertions.Execution.AssertionScope.FailWith(String message, Object[] args)
   at FluentAssertions.Collections.CollectionAssertions`2.NotBeEmpty(String because, Object[] becauseArgs)
   at Elastic.Apm.Tests.BackendCommTests.CentralConfig.CentralConfigFetcherTests.Should_Update_IgnoreMessageQueues_Configuration()
Standard Output:
    Elastic APM .NET Tests> 312> Xunit> [2021-06-23 23:58:53.029 +00:00][Info] - {LoggingTestBase} Starting test: Elastic.Apm.Tests.BackendCommTests.CentralConfig.CentralConfigFetcherTests.Should_Update_IgnoreMessageQueues_Configuration...
    Elastic APM .NET Tests> 312> Xunit> [2021-06-23 23:58:53.041 +00:00][Error] - {CentralConfigFetcher} Exception was thrown while fetching configuration from APM Server and parsing it. ETag: `<null>'. URL: `http://localhost:8200/config/v1/agents?service.name=xunit_execution_desktop'. Apm Server base URL: `http://localhost:8200/'. WaitInterval: 5m. dbgIterationsCount: 1.
    Elastic APM .NET Tests> 312> Xunit> +-> Request:
    Elastic APM .NET Tests> 312> Xunit>     Method: GET, RequestUri: 'http://localhost:8200/config/v1/agents?service.name=xunit_execution_desktop', Version: 1.1, Content: <null>, Headers:
    Elastic APM .NET Tests> 312> Xunit>     {
    Elastic APM .NET Tests> 312> Xunit>       User-Agent: elasticapm-dotnet/1.10.0+520d06f38e81f908316329612d6d08a058c36964
    Elastic APM .NET Tests> 312> Xunit>       User-Agent: System.Net.Http/4.8.3761.0
    Elastic APM .NET Tests> 312> Xunit>       User-Agent: .NET_Framework/4.8.4300.0
    Elastic APM .NET Tests> 312> Xunit>     }
    Elastic APM .NET Tests> 312> Xunit> +-> Response:
    Elastic APM .NET Tests> 312> Xunit>     StatusCode: 200, ReasonPhrase: 'OK', Version: 1.1, Content: System.Net.Http.StringContent, Headers:
    Elastic APM .NET Tests> 312> Xunit>     {
    Elastic APM .NET Tests> 312> Xunit>       ETag: "etag"
    Elastic APM .NET Tests> 312> Xunit>       Content-Type: text/plain; charset=utf-8
    Elastic APM .NET Tests> 312> Xunit>       Content-Length: 34
    Elastic APM .NET Tests> 312> Xunit>     }
    Elastic APM .NET Tests> 312> Xunit> +-> Response body [length: 34]:
    Elastic APM .NET Tests> 312> Xunit>     { "ignore_message_queues": "foo" }
    Elastic APM .NET Tests> 312> Xunit> +-> Exception: System.NullReferenceException: Object reference not set to an instance of an object.
    Elastic APM .NET Tests> 312> Xunit>    at Elastic.Apm.BackendComm.CentralConfig.CentralConfigFetcher.<WorkLoopIteration>d__15.MoveNext() in C:\Users\jenkins\workspace\net_apm-agent-dotnet-mbp_PR-1271\apm-agent-dotnet\src\Elastic.Apm\BackendComm\CentralConfig\CentralConfigFetcher.cs:line 107
    Elastic APM .NET Tests> 312> Xunit> [2021-06-23 23:59:53.128 +00:00][Info] - {LoggingTestBase} Finished test: Elastic.Apm.Tests.BackendCommTests.CentralConfig.CentralConfigFetcherTests.Should_Update_IgnoreMessageQueues_Configuration 
 ```
</p></details>
</ul>
    
##### `Initializing / Parallel / Linux / Test / CentralConfigFetcherTests.Elastic.Apm.Tests.BackendCommTests.CentralConfig.CentralConfigFetcherTests.Should_Update_Logger_That_Is_ILogLevelSwitchable – Elastic.Apm.Tests.BackendCommTests.CentralConfig.CentralConfigFetcherTests`


<ul>
<details><summary>Expand to view the error details</summary><p>

```
 Expected testLogger.LogLevelSwitch.Level to be Error, but found Trace. 
 ```
</p></details>
<details><summary>Expand to view the stacktrace</summary><p>

```
 Expected testLogger.LogLevelSwitch.Level to be Error, but found Trace.
Stack Trace:
   at FluentAssertions.Execution.XUnit2TestFramework.Throw(String message)
   at FluentAssertions.Execution.TestFrameworkProvider.Throw(String message)
   at FluentAssertions.Execution.DefaultAssertionStrategy.HandleFailure(String message)
   at FluentAssertions.Execution.AssertionScope.FailWith(Func`1 failReasonFunc)
   at FluentAssertions.Primitives.ObjectAssertions.Be(Object expected, String because, Object[] becauseArgs)
   at Elastic.Apm.Tests.BackendCommTests.CentralConfig.CentralConfigFetcherTests.Should_Update_Logger_That_Is_ILogLevelSwitchable() in /var/lib/jenkins/workspace/net_apm-agent-dotnet-mbp_PR-1271/apm-agent-dotnet/test/Elastic.Apm.Tests/BackendCommTests/CentralConfig/CentralConfigFetcherTests.cs:line 78
Standard Output:
    Elastic APM .NET Tests> 324> Xunit> [2021-06-23 23:53:39.690 +00:00][Info] - {LoggingTestBase} Starting test: Elastic.Apm.Tests.BackendCommTests.CentralConfig.CentralConfigFetcherTests.Should_Update_Logger_That_Is_ILogLevelSwitchable...
    Elastic APM .NET Tests> 324> Xunit> [2021-06-23 23:54:39.939 +00:00][Info] - {LoggingTestBase} Finished test: Elastic.Apm.Tests.BackendCommTests.CentralConfig.CentralConfigFetcherTests.Should_Update_Logger_That_Is_ILogLevelSwitchable 
 ```
</p></details>
</ul>
    
</p></details>


<!-- STEPS ERRORS IF ANY -->









### Steps errors [![3](https://img.shields.io/badge/3%20-red)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-dotnet%2Fapm-agent-dotnet-mbp%2FPR-1271/detail/PR-1271/51//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>


  
  
  
  
##### `Test & coverage`
<ul>
<li>Took 2 min 28 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-dotnet/pipelines/apm-agent-dotnet-mbp/pipelines/PR-1271/runs/51/steps/454/log/?start=0">here</a></li>
<li>Description: <code>.ci/windows/testnet461.bat</code></l1>
</ul>
  
  
  
  
##### `Archive the artifacts`
<ul>
<li>Took 0 min 0 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-dotnet/pipelines/apm-agent-dotnet-mbp/pipelines/PR-1271/runs/51/steps/501/log/?start=0">here</a></li>
<li>Description: <code>[2021-06-24T00:00:01.387Z] Archiving artifacts
script returned exit code 1</code></l1>
</ul>
  
  
  
  
##### `Build`
<ul>
<li>Took 1 min 35 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-dotnet/pipelines/apm-agent-dotnet-mbp/pipelines/PR-1271/runs/51/steps/348/log/?start=0">here</a></li>
<li>Description: <code>.ci/windows/dotnet.bat</code></l1>
</ul>
  
  </p>
  </details>



<!-- LOG OUTPUT-->
### Log output

<details><summary>Expand to view the last 100 lines of log output</summary>
<p>

```
[2021-06-24T00:21:45.725Z] Copied Elastic.Apm/Libraries/Newtonsoft.Json/Schema/ValidationEventArgs.cs. 
[2021-06-24T00:21:45.725Z] Starting copy source file Elastic.Apm/Libraries/Newtonsoft.Json/Converters/StringEnumConverter.cs. 
[2021-06-24T00:21:45.800Z] Copied Elastic.Apm/Libraries/Newtonsoft.Json/Converters/StringEnumConverter.cs. 
[2021-06-24T00:21:45.800Z] Starting copy source file Elastic.Apm/Libraries/Newtonsoft.Json/Utilities/DynamicProxyMetaObject.cs. 
[2021-06-24T00:21:45.875Z] Copied Elastic.Apm/Libraries/Newtonsoft.Json/Utilities/DynamicProxyMetaObject.cs. 
[2021-06-24T00:21:45.875Z] Starting copy source file Libraries/Newtonsoft.Json/Converters/StringEnumConverter.cs. 
[2021-06-24T00:21:45.955Z] Copied Libraries/Newtonsoft.Json/Converters/StringEnumConverter.cs. 
[2021-06-24T00:21:45.955Z] Starting copy source file Libraries/Newtonsoft.Json/Serialization/JsonPrimitiveContract.cs. 
[2021-06-24T00:21:46.024Z] Copied Libraries/Newtonsoft.Json/Serialization/JsonPrimitiveContract.cs. 
[2021-06-24T00:21:46.024Z] Starting copy source file Elastic.Apm/Config/AbstractConfigurationReader.cs. 
[2021-06-24T00:21:46.110Z] Copied Elastic.Apm/Config/AbstractConfigurationReader.cs. 
[2021-06-24T00:21:46.111Z] Starting copy source file Elastic.Apm/Filters/SpanStackTraceCapturingFilter.cs. 
[2021-06-24T00:21:46.190Z] Copied Elastic.Apm/Filters/SpanStackTraceCapturingFilter.cs. 
[2021-06-24T00:21:46.190Z] Starting copy source file Libraries/Newtonsoft.Json/Schema/JsonSchemaBuilder.cs. 
[2021-06-24T00:21:46.267Z] Copied Libraries/Newtonsoft.Json/Schema/JsonSchemaBuilder.cs. 
[2021-06-24T00:21:46.267Z] Starting copy source file Elastic.Apm/Helpers/ContractExtensions.cs. 
[2021-06-24T00:21:46.341Z] Copied Elastic.Apm/Helpers/ContractExtensions.cs. 
[2021-06-24T00:21:46.341Z] Starting copy source file Report/Serialization/MetricSetConverter.cs. 
[2021-06-24T00:21:46.416Z] Copied Report/Serialization/MetricSetConverter.cs. 
[2021-06-24T00:21:46.416Z] Starting copy source file Libraries/Newtonsoft.Json/Utilities/DynamicProxy.cs. 
[2021-06-24T00:21:46.489Z] Copied Libraries/Newtonsoft.Json/Utilities/DynamicProxy.cs. 
[2021-06-24T00:21:46.489Z] Starting copy source file Cloud/AzureCloudMetadataProvider.cs. 
[2021-06-24T00:21:46.561Z] Copied Cloud/AzureCloudMetadataProvider.cs. 
[2021-06-24T00:21:46.561Z] Starting copy source file Api/Context.cs. 
[2021-06-24T00:21:46.637Z] Copied Api/Context.cs. 
[2021-06-24T00:21:46.638Z] Starting copy source file Elastic.Apm/Libraries/Newtonsoft.Json/Serialization/DefaultContractResolver.cs. 
[2021-06-24T00:21:46.726Z] Copied Elastic.Apm/Libraries/Newtonsoft.Json/Serialization/DefaultContractResolver.cs. 
[2021-06-24T00:21:46.726Z] Starting copy source file Config/AbstractConfigurationReader.cs. 
[2021-06-24T00:21:46.811Z] Copied Config/AbstractConfigurationReader.cs. 
[2021-06-24T00:21:46.811Z] Starting copy source file Elastic.Apm/Logging/LogValuesFormatter.cs. 
[2021-06-24T00:21:46.886Z] Copied Elastic.Apm/Logging/LogValuesFormatter.cs. 
[2021-06-24T00:22:20.815Z] [Checks API] No suitable checks publisher found.
[2021-06-24T00:22:20.909Z] [INFO] getVaultSecret: Getting secrets
[2021-06-24T00:22:21.141Z] Masking supported pattern matches of $VAULT_ADDR or $VAULT_ROLE_ID or $VAULT_SECRET_ID
[2021-06-24T00:22:21.764Z] Running in /var/lib/jenkins/workspace/net_apm-agent-dotnet-mbp_PR-1271/apm-agent-dotnet
[2021-06-24T00:22:21.783Z] [INFO] Codecov: Getting branch ref...
[2021-06-24T00:22:21.858Z] Masking supported pattern matches of $GITHUB_TOKEN
[2021-06-24T00:22:22.298Z] [INFO] Codecov: Sending data...
[2021-06-24T00:22:22.693Z] + echo curl -sSLo codecov.sh https://codecov.io/bash
[2021-06-24T00:22:22.694Z] curl -sSLo codecov.sh https://codecov.io/bash
[2021-06-24T00:22:23.010Z] + echo bash codecov.sh
[2021-06-24T00:22:23.010Z] bash codecov.sh
[2021-06-24T00:22:23.111Z] Archiving artifacts
[2021-06-24T00:22:23.211Z] java.lang.InterruptedException: no matches found within 10000
[2021-06-24T00:22:23.211Z] 	at hudson.FilePath$ValidateAntFileMask.hasMatch(FilePath.java:3069)
[2021-06-24T00:22:23.211Z] 	at hudson.FilePath$ValidateAntFileMask.invoke(FilePath.java:2978)
[2021-06-24T00:22:23.211Z] 	at hudson.FilePath$ValidateAntFileMask.invoke(FilePath.java:2929)
[2021-06-24T00:22:23.211Z] 	at hudson.FilePath$FileCallableWrapper.call(FilePath.java:3317)
[2021-06-24T00:22:23.211Z] Also:   hudson.remoting.Channel$CallSiteStackTrace: Remote call to JNLP4-connect connection from apm-ci-immutable-ubuntu-1804-1624491206901416871.c.elastic-ci-prod.internal/10.224.0.143:45152
[2021-06-24T00:22:23.211Z] 		at hudson.remoting.Channel.attachCallSiteStackTrace(Channel.java:1800)
[2021-06-24T00:22:23.211Z] 		at hudson.remoting.UserRequest$ExceptionResponse.retrieve(UserRequest.java:356)
[2021-06-24T00:22:23.211Z] 		at hudson.remoting.Channel.call(Channel.java:1001)
[2021-06-24T00:22:23.211Z] 		at hudson.FilePath.act(FilePath.java:1159)
[2021-06-24T00:22:23.211Z] 		at hudson.FilePath.act(FilePath.java:1148)
[2021-06-24T00:22:23.211Z] 		at hudson.FilePath.validateAntFileMask(FilePath.java:2927)
[2021-06-24T00:22:23.211Z] 		at hudson.tasks.ArtifactArchiver.perform(ArtifactArchiver.java:270)
[2021-06-24T00:22:23.211Z] 		at org.jenkinsci.plugins.workflow.steps.CoreStep$Execution.run(CoreStep.java:99)
[2021-06-24T00:22:23.211Z] 		at org.jenkinsci.plugins.workflow.steps.CoreStep$Execution.run(CoreStep.java:69)
[2021-06-24T00:22:23.211Z] 		at org.jenkinsci.plugins.workflow.steps.SynchronousNonBlockingStepExecution.lambda$start$0(SynchronousNonBlockingStepExecution.java:47)
[2021-06-24T00:22:23.211Z] 		at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
[2021-06-24T00:22:23.211Z] 		at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[2021-06-24T00:22:23.211Z] 		at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[2021-06-24T00:22:23.211Z] 		at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[2021-06-24T00:22:23.211Z] Caused: hudson.FilePath$TunneledInterruptedException
[2021-06-24T00:22:23.211Z] 	at hudson.FilePath$FileCallableWrapper.call(FilePath.java:3319)
[2021-06-24T00:22:23.211Z] 	at hudson.remoting.UserRequest.perform(UserRequest.java:211)
[2021-06-24T00:22:23.211Z] 	at hudson.remoting.UserRequest.perform(UserRequest.java:54)
[2021-06-24T00:22:23.211Z] 	at hudson.remoting.Request$2.run(Request.java:375)
[2021-06-24T00:22:23.211Z] 	at hudson.remoting.InterceptingExecutorService$1.call(InterceptingExecutorService.java:73)
[2021-06-24T00:22:23.211Z] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[2021-06-24T00:22:23.212Z] 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[2021-06-24T00:22:23.212Z] 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[2021-06-24T00:22:23.212Z] 	at hudson.remoting.Engine$1.lambda$newThread$0(Engine.java:118)
[2021-06-24T00:22:23.212Z] Caused: java.lang.InterruptedException: java.lang.InterruptedException: no matches found within 10000
[2021-06-24T00:22:23.212Z] 	at hudson.FilePath.act(FilePath.java:1161)
[2021-06-24T00:22:23.212Z] 	at hudson.FilePath.act(FilePath.java:1148)
[2021-06-24T00:22:23.212Z] 	at hudson.FilePath.validateAntFileMask(FilePath.java:2927)
[2021-06-24T00:22:23.212Z] 	at hudson.tasks.ArtifactArchiver.perform(ArtifactArchiver.java:270)
[2021-06-24T00:22:23.212Z] 	at org.jenkinsci.plugins.workflow.steps.CoreStep$Execution.run(CoreStep.java:99)
[2021-06-24T00:22:23.212Z] 	at org.jenkinsci.plugins.workflow.steps.CoreStep$Execution.run(CoreStep.java:69)
[2021-06-24T00:22:23.212Z] 	at org.jenkinsci.plugins.workflow.steps.SynchronousNonBlockingStepExecution.lambda$start$0(SynchronousNonBlockingStepExecution.java:47)
[2021-06-24T00:22:23.212Z] 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
[2021-06-24T00:22:23.212Z] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[2021-06-24T00:22:23.212Z] 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[2021-06-24T00:22:23.212Z] 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[2021-06-24T00:22:23.212Z] 	at java.lang.Thread.run(Thread.java:748)
[2021-06-24T00:22:23.212Z] No artifacts found that match the file pattern "/var/lib/jenkins/workspace/net_apm-agent-dotnet-mbp_PR-1271/**/MSBuild_*.failure.txt". Configuration error?
[2021-06-24T00:22:23.385Z] Stage "Release to feedz.io" skipped due to earlier failure(s)
[2021-06-24T00:22:23.425Z] Stage "Release" skipped due to earlier failure(s)
[2021-06-24T00:22:23.442Z] Stage "Release" skipped due to earlier failure(s)
[2021-06-24T00:22:23.483Z] Stage "Release" skipped due to earlier failure(s)
[2021-06-24T00:22:23.546Z] Stage "AfterRelease" skipped due to earlier failure(s)
[2021-06-24T00:22:23.563Z] Stage "AfterRelease" skipped due to earlier failure(s)
[2021-06-24T00:22:23.785Z] Running on Jenkins in /var/lib/jenkins/workspace/net_apm-agent-dotnet-mbp_PR-1271
[2021-06-24T00:22:23.837Z] [INFO] getVaultSecret: Getting secrets
[2021-06-24T00:22:23.879Z] Masking supported pattern matches of $VAULT_ADDR or $VAULT_ROLE_ID or $VAULT_SECRET_ID
[2021-06-24T00:22:24.554Z] + chmod 755 generate-build-data.sh
[2021-06-24T00:22:24.554Z] + ./generate-build-data.sh https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-dotnet/apm-agent-dotnet-mbp/PR-1271/ https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-dotnet/apm-agent-dotnet-mbp/PR-1271/runs/51 FAILURE 2991060
[2021-06-24T00:22:24.554Z] INFO: curl https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-dotnet/apm-agent-dotnet-mbp/PR-1271/runs/51/steps/?limit=10000 -o steps-info.json
[2021-06-24T00:22:25.256Z] INFO: curl https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-dotnet/apm-agent-dotnet-mbp/PR-1271/runs/51/tests/?status=FAILED -o tests-errors.json

```

</p>
</details>



<!--COMMENT_GENERATED_WITH_ID_comment.id-->
